langchain/src/retrievers/supabase-hybrid-search.ts

```diff
@@ -0,0 +1,177 @@
+  keywordK?: number;
+}
+
+export interface SupabaseHybridSearchParams {
+  query: string;
+  similarityK: number;
+  keywordK: number;
+}
+
+export class SupabaseHybridSearch extends BaseRetriever {
+  similarityK: number;
+
+  query: string;
+
+  keywordK: number;
+
+  similarityQueryName: string;
+
+  client: SupabaseClient;
+
+  tableName: string;
+
+  keywordQueryName: string;
+
+  embeddings: Embeddings;
+
+  constructor(embeddings: Embeddings, args: SupabaseLibArgs) {
+    super();
+    this.embeddings = embeddings;
+    this.client = args.client;
+    this.tableName = args.tableName || "documents";
+    this.similarityQueryName = args.similarityQueryName || "match_documents";
+    this.keywordQueryName = args.keywordQueryName || "kw_match_documents";
+    this.similarityK = args.similarityK || 2;
+    this.keywordK = args.keywordK || 2;
+  }
+
+  protected async similaritySearch(
+    query: string,
+    k: number
+  ): Promise<[Document, number][]> {
+    const embeddedQuery = await this.embeddings.embedQuery(query);
+
+    const matchDocumentsParams: SearchEmbeddingsParams = {
+      query_embedding: embeddedQuery,
+      match_count: k,
+    };
+
+    const { data: searches, error } = await this.client.rpc(
+      this.similarityQueryName,
+      matchDocumentsParams
+    );
+
+    if (error) {
+      throw new Error(`Error searching for documents: ${error}`);
+    }
+
+    const result: [Document, number][] = (
+      searches as SearchEmbeddingsResponse[]
+    ).map((resp) => [
+      new Document({
+        metadata: resp.metadata,
+        pageContent: resp.content,
+      }),
+      resp.similarity,
+    ]);
+    return result;
+  }
+
+  protected async keywordSearch(
+    query: string,
+    k: number
+  ): Promise<[Document, number][]> {
+    const kwMatchDocumentsParams: kwSearchParams = {
+      query_text: query,
+      match_count: k,
+    };
+
+    const { data: searches, error } = await this.client.rpc(
+      this.keywordQueryName,
+      kwMatchDocumentsParams
+    );
+
+    if (error) {
+      throw new Error(`Error searching for documents: ${error}`);
+    }
+
+    const result: [Document, number][] = (searches as kwSearchResponse[]).map(
+      (resp) => [
+        new Document({
+          metadata: resp.metadata,
+          pageContent: resp.content,
+        }),
+        resp.similarity * 10,
+      ]
+    );
+    return result;
+  }
+
+  protected async hybridSearch(
+    query: string,
+    similarityK: number,
+    keywordK: number
+  ): Promise<[Document, number][]> {
+    const similarity_search = this.similaritySearch(query, similarityK);
+
+    const keyword_search = this.keywordSearch(query, keywordK);
+
+    return Promise.all<[Document, number][]>([
+      similarity_search,
+      keyword_search,
+    ])
+      .then((results) => results.flat())
+      .then((results) => {
+        const seenContent = new Set();
+        return results.filter(([doc]) => {
+          const content = doc.pageContent;
```

Comment:
@ericciarla one last comment, I think this should either take the max or the sum of the scores for documents found in both searches, thoughts?

Reply:
@nfcampos I am alright with that. I think max score would be best. Is that some you want to quickly add? I can as well, just let me know